crates/wit-component/src/encoding/wit/v2.rs

```diff
@@ -57,33 +51,101 @@ impl Encoder<'_> {
         // decoding process where everyone's view of a foreign document agrees
         // notably on the order that types are defined in to assist with
         // roundtripping.
-        for (name, &id) in self.resolve.packages[self.package].interfaces.iter() {
-            let component_ty = self.encode_interface(id)?;
-            let ty = self.component.type_component(&component_ty);
-            self.component
-                .export(name.as_ref(), ComponentExportKind::Type, ty, None);
+        let mut names = NameMap::new();
+        for pkg in self.packages {
+            if self.packages.len() > 1 {
+                let sub_encoder = Encoder {
+                    component: ComponentBuilder::default(),
+                    resolve: self.resolve,
+                    packages: self.packages,
+                };
+                self.encode_package(&mut names, pkg, Some(sub_encoder))?;
+            } else {
+                self.encode_package(&mut names, pkg, None)?;
+            }
+        }
+        if self.packages.len() == 1 {
+            // let pkg = self.packages[0];
+            let package_metadata = PackageMetadata::extract(self.resolve, self.packages[0]);
+            self.component.custom_section(&CustomSection {
+                name: PackageMetadata::SECTION_NAME.into(),
+                data: package_metadata.encode()?.into(),
+            });
         }
+        let mut final_names = ComponentNameSection::new();
+        final_names.components(&names);
+        self.component.names(&final_names);
 
-        for (name, &world) in self.resolve.packages[self.package].worlds.iter() {
-            // Encode the `world` directly as a component, then create a wrapper
-            // component that exports that component.
-            let component_ty = super::encode_world(self.resolve, world)?;
+        Ok(())
+    }
 
-            let world = &self.resolve.worlds[world];
-            let mut wrapper = ComponentType::new();
-            wrapper.ty().component(&component_ty);
-            let pkg = &self.resolve.packages[world.package.unwrap()];
-            wrapper.export(&pkg.name.interface_id(name), ComponentTypeRef::Component(0));
+    fn encode_package(
+        &mut self,
+        names: &mut NameMap,
+        pkg: &PackageId,
+        mut sub_encoder: Option<Encoder>,
```

Comment:
Instead of passing around `sub_encoder` and being careful to use-it-or-not, could the `self.component` field be `mem::replace`'d to sort of swap out where the encoding is happening within?

Reply:
(it would also help the duplication below for each `if let` with the `sub_encoder`)